File: calculator.py
"""
Functional Calculator - Pure functional implementation in Python
No classes, no global state, no mutations
"""

from functools import reduce
from typing import Union, List, Tuple


def tokenize(expr: str) -> Tuple:
    """
    Parse an arithmetic expression string into tokens.
    
    Args:
        expr: String expression
        
    Returns:
        Tuple of tokens (numbers, operators, parentheses)
        
    Raises:
        ValueError: If expression is invalid
    """
    expr_clean = expr.replace(" ", "")
    
    if not expr_clean:
        raise ValueError("Malformed expression")
    
    tokens = []
    i = 0
    
    while i < len(expr_clean):
        char = expr_clean[i]
        
        # Handle numbers (integers and floats)
        if char.isdigit() or (char == '.' and i + 1 < len(expr_clean) and expr_clean[i + 1].isdigit()):
            j = i
            has_dot = False
            
            while j < len(expr_clean) and (expr_clean[j].isdigit() or (expr_clean[j] == '.' and not has_dot)):
                if expr_clean[j] == '.':
                    has_dot = True
                j += 1
            
            tokens.append(float(expr_clean[i:j]))
            i = j
        
        # Handle operators and parentheses
        elif char in "+-*/()":
            tokens.append(char)
            i += 1
        
        else:
            raise ValueError("Invalid character")
    
    return tuple(tokens)


def process_unary_minus(tokens: Tuple) -> Tuple:
    """
    Convert minus operators to unary minus where appropriate.
    
    Args:
        tokens: Tuple of tokens
        
    Returns:
        Tuple with unary minus marked as 'NEG'
    """
    result = []
    
    for i, token in enumerate(tokens):
        if token == '-':
            # Check if it's a unary minus
            prev_token = tokens[i - 1] if i > 0 else None
            is_unary = i == 0 or prev_token in ('(', '+', '-', '*', '/')
            
            result.append('NEG' if is_unary else token)
        else:
            result.append(token)
    
    return tuple(result)


def get_precedence(op: str) -> int:
    """Get operator precedence."""
    if op == 'NEG':
        return 3
    elif op in ('+', '-'):
        return 1
    elif op in ('*', '/'):
        return 2
    return 0


def is_operator(token) -> bool:
    """Check if token is an operator."""
    return token in ('+', '-', '*', '/', 'NEG')


def shunting_yard(tokens: Tuple) -> Tuple:
    """
    Convert infix notation to postfix (RPN) using Shunting-yard algorithm.
    
    Args:
        tokens: Tuple of tokens in infix notation
        
    Returns:
        Tuple of tokens in RPN
        
    Raises:
        ValueError: If parentheses are mismatched
    """
    output_queue = []
    operator_stack = []
    
    for token in tokens:
        # Number
        if isinstance(token, float):
            output_queue.append(token)
        
        # Operator
        elif is_operator(token):
            while (operator_stack and 
                   operator_stack[-1] != '(' and 
                   is_operator(operator_stack[-1]) and
                   get_precedence(operator_stack[-1]) >= get_precedence(token)):
                output_queue.append(operator_stack.pop())
            operator_stack.append(token)
        
        # Left parenthesis
        elif token == '(':
            operator_stack.append(token)
        
        # Right parenthesis
        elif token == ')':
            found_left = False
            while operator_stack:
                op = operator_stack.pop()
                if op == '(':
                    found_left = True
                    break
                output_queue.append(op)
            
            if not found_left:
                raise ValueError("Mismatched parentheses")
    
    # Pop remaining operators
    while operator_stack:
        op = operator_stack.pop()
        if op in ('(', ')'):
            raise ValueError("Mismatched parentheses")
        output_queue.append(op)
    
    return tuple(output_queue)


def to_rpn(tokens: Tuple) -> Tuple:
    """
    Convert tokens to Reverse Polish Notation.
    
    Args:
        tokens: Tuple of tokens
        
    Returns:
        Tuple of tokens in RPN
    """
    processed_tokens = process_unary_minus(tokens)
    return shunting_yard(processed_tokens)


def apply_operator(op: str, left: float, right: float = None) -> float:
    """
    Apply an operator to operand(s).
    
    Args:
        op: Operator string
        left: Left operand (or only operand for unary operators)
        right: Right operand (optional for unary operators)
        
    Returns:
        Result of the operation
        
    Raises:
        ValueError: For division by zero or invalid operators
    """
    if op == '+':
        return left + right
    elif op == '-':
        return left - right
    elif op == '*':
        return left * right
    elif op == '/':
        if right == 0:
            raise ValueError("Division by zero")
        return left / right
    elif op == 'NEG':
        return -left
    else:
        raise ValueError("Malformed expression")


def eval_rpn(rpn: Tuple) -> float:
    """
    Evaluate an RPN expression.
    
    Args:
        rpn: Tuple of tokens in RPN
        
    Returns:
        Result of the evaluation
        
    Raises:
        ValueError: If RPN is malformed
    """
    stack = []
    
    for token in rpn:
        if isinstance(token, float):
            stack.append(token)
        elif is_operator(token):
            if token == 'NEG':
                if len(stack) < 1:
                    raise ValueError("Malformed expression")
                operand = stack.pop()
                result = apply_operator(token, operand)
                stack.append(result)
            else:
                if len(stack) < 2:
                    raise ValueError("Malformed expression")
                right = stack.pop()
                left = stack.pop()
                result = apply_operator(token, left, right)
                stack.append(result)
    
    if len(stack) != 1:
        raise ValueError("Malformed expression")
    
    return stack[0]


def format_result(x: float) -> str:
    """
    Format the result according to rules:
    - If integer: print without decimal point (e.g., 7)
    - If fractional: print as decimal (e.g., 2.5)
    
    Args:
        x: Floating point number
        
    Returns:
        Formatted string
    """
    if x == int(x):
        return str(int(x))
    else:
        return str(x)


def calculate(expr: str) -> str:
    """
    Main calculation function.
    Takes an expression string and returns the formatted result or error message.
    
    Args:
        expr: String expression
        
    Returns:
        Result string or ERROR message
    """
    try:
        tokens = tokenize(expr)
        rpn = to_rpn(tokens)
        result = eval_rpn(rpn)
        return format_result(result)
    except ValueError as e:
        return f"ERROR: {str(e)}"


def main():
    """
    Main entry point - reads from console and prints result.
    """
    expr = input()
    print(calculate(expr))


if __name__ == "__main__":
    main()


if __name__ == "__main__":
    main()

File: calculator.js
// Tokenizer function - parses expression into tokens
function tokenize(expr) {
    const exprClean = expr.replace(/\s+/g, "");
    
    if (!exprClean) {
        throw new Error("Malformed expression");
    }
    
    const tokens = [];
    let i = 0;
    
    while (i < exprClean.length) {
        const char = exprClean[i];
        
        // Handle numbers (integers and floats)
        if (/\d/.test(char) || (char === '.' && i + 1 < exprClean.length && /\d/.test(exprClean[i + 1]))) {
            let j = i;
            let hasDot = false;
            
            while (j < exprClean.length && (/\d/.test(exprClean[j]) || (exprClean[j] === '.' && !hasDot))) {
                if (exprClean[j] === '.') {
                    hasDot = true;
                }
                j++;
            }
            
            tokens.push(parseFloat(exprClean.substring(i, j)));
            i = j;
        }
        // Handle operators and parentheses
        else if ("+-*/()".includes(char)) {
            tokens.push(char);
            i++;
        }
        else {
            throw new Error("Invalid character");
        }
    }
    
    return tokens;
}

// Handle unary minus
function handleUnaryMinus(tokens) {
    const result = [];
    
    for (let i = 0; i < tokens.length; i++) {
        const token = tokens[i];
        
        if (token === '-') {
            // Check if it's a unary minus
            const prevToken = i > 0 ? tokens[i - 1] : null;
            if (i === 0 || prevToken === '(' || prevToken === '+' || prevToken === '-' || prevToken === '*' || prevToken === '/') {
                result.push('NEG');
            } else {
                result.push(token);
            }
        } else {
            result.push(token);
        }
    }
    
    return result;
}

// Get operator precedence
function precedence(op) {
    if (['+', '-'].includes(op)) return 1;
    if (['*', '/'].includes(op)) return 2;
    if (op === 'NEG') return 3;
    return 0;
}

// Check if token is an operator
function isOperator(token) {
    return ['+', '-', '*', '/', 'NEG'].includes(token);
}

// Shunting-yard algorithm
function shuntingYard(tokens) {
    const outputQueue = [];
    const operatorStack = [];
    
    for (const token of tokens) {
        // Number
        if (typeof token === 'number') {
            outputQueue.push(token);
        }
        // Operator
        else if (isOperator(token)) {
            while (operatorStack.length > 0 &&
                   operatorStack[operatorStack.length - 1] !== '(' &&
                   isOperator(operatorStack[operatorStack.length - 1]) &&
                   precedence(operatorStack[operatorStack.length - 1]) >= precedence(token)) {
                outputQueue.push(operatorStack.pop());
            }
            operatorStack.push(token);
        }
        // Left parenthesis
        else if (token === '(') {
            operatorStack.push(token);
        }
        // Right parenthesis
        else if (token === ')') {
            let foundLeft = false;
            
            while (operatorStack.length > 0) {
                const op = operatorStack.pop();
                if (op === '(') {
                    foundLeft = true;
                    break;
                }
                outputQueue.push(op);
            }
            
            if (!foundLeft) {
                throw new Error("Mismatched parentheses");
            }
        }
    }
    
    // Pop remaining operators
    while (operatorStack.length > 0) {
        const op = operatorStack.pop();
        if (['(', ')'].includes(op)) {
            throw new Error("Mismatched parentheses");
        }
        outputQueue.push(op);
    }
    
    return outputQueue;
}

// Convert to RPN
function toRPN(tokens) {
    const processedTokens = handleUnaryMinus(tokens);
    return shuntingYard(processedTokens);
}

// Apply operator
function applyOperator(op, operands) {
    switch (op) {
        case '+':
            return operands[1] + operands[0];
        case '-':
            return operands[1] - operands[0];
        case '*':
            return operands[1] * operands[0];
        case '/':
            if (operands[0] === 0) {
                throw new Error("Division by zero");
            }
            return operands[1] / operands[0];
        case 'NEG':
            return -operands[0];
        default:
            throw new Error("Malformed expression");
    }
}

// Evaluate RPN
function evalRPN(rpn) {
    const stack = [];
    
    for (const token of rpn) {
        if (typeof token === 'number') {
            stack.push(token);
        } else if (isOperator(token)) {
            if (token === 'NEG') {
                if (stack.length < 1) {
                    throw new Error("Malformed expression");
                }
                const operand = stack.pop();
                const result = applyOperator(token, [operand]);
                stack.push(result);
            } else {
                if (stack.length < 2) {
                    throw new Error("Malformed expression");
                }
                const right = stack.pop();
                const left = stack.pop();
                const result = applyOperator(token, [right, left]);
                stack.push(result);
            }
        }
    }
    
    if (stack.length !== 1) {
        throw new Error("Malformed expression");
    }
    
    return stack[0];
}

// Format result
function formatResult(x) {
    // Check if the result is an integer value
    if (x === Math.floor(x)) {
        return String(Math.floor(x));
    } else {
        return String(x);
    }
}

// Main calculation function
function calculate(expr) {
    try {
        const tokens = tokenize(expr);
        const rpn = toRPN(tokens);
        const result = evalRPN(rpn);
        return formatResult(result);
    } catch (e) {
        return `ERROR: ${e.message}`;
    }
}

// DOM manipulation
document.addEventListener('DOMContentLoaded', function() {
    const expressionInput = document.getElementById('expression');
    const calcBtn = document.getElementById('calcBtn');
    const clearBtn = document.getElementById('clearBtn');
    const resultDiv = document.getElementById('result');
    
    function performCalculation() {
        const expr = expressionInput.value.trim();
        
        if (!expr) {
            resultDiv.textContent = '';
            resultDiv.classList.remove('success', 'error', 'empty');
            return;
        }
        
        const result = calculate(expr);
        
        if (result.startsWith('ERROR')) {
            resultDiv.textContent = result;
            resultDiv.classList.remove('success', 'empty');
            resultDiv.classList.add('error');
        } else {
            resultDiv.textContent = `Result: ${result}`;
            resultDiv.classList.remove('error', 'empty');
            resultDiv.classList.add('success');
        }
    }
    
    calcBtn.addEventListener('click', performCalculation);
    
    // Allow Enter key to calculate
    expressionInput.addEventListener('keypress', function(event) {
        if (event.key === 'Enter') {
            performCalculation();
        }
    });
    
    clearBtn.addEventListener('click', function() {
        expressionInput.value = '';
        expressionInput.focus();
        resultDiv.textContent = '';
        resultDiv.classList.remove('success', 'error', 'empty');
    });
    
    expressionInput.focus();
});
